sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+        cache.mark_endpoint_unavailable(unavailable_endpoint, operation);
+
+        // check that endpoint is last option in read endpoints and it is in the location unavailability info map
+        assert_eq!(
+            cache.locations_info.read_endpoints.last(),
+            Some(&unavailable_endpoint.to_string())
+        );
+
+        assert!(cache.is_endpoint_unavailable(unavailable_endpoint, operation));
+    }
+
+    #[test]
+    fn mark_write_endpoint_unavailable_test() {
+        // set up test data
+        let (default_endpoint, write_locations, read_locations, preferred_locations) =
+            create_test_data();
+
+        let mut cache = LocationCache::new(default_endpoint, preferred_locations);
+        cache.update(write_locations, read_locations);
+
+        // mark location 1 as unavailable endpoint for write operation
+        let unavailable_endpoint = "https://location1.documents.azure.com";
+        let operation = RequestOperation::Write;
+        cache.mark_endpoint_unavailable(unavailable_endpoint, operation);
+
+        // check that endpoint is last option in write endpoints and it is in the location unavailability info map
+        assert_eq!(
+            cache.locations_info.write_endpoints.last(),
+            Some(&unavailable_endpoint.to_string())
+        );
+
+        assert!(cache.is_endpoint_unavailable(unavailable_endpoint, operation));
+    }
+
+    #[test]
+    fn mark_same_endpoint_unavailable_test() {
+        let (default_endpoint, write_locations, read_locations, preferred_locations) =
+            create_test_data();
+
+        let mut cache = LocationCache::new(default_endpoint, preferred_locations);
+
+        cache.update(write_locations, read_locations);
+
+        let endpoint1 = "https://location1.documents.azure.com";
+
+        cache.mark_endpoint_unavailable(endpoint1, RequestOperation::Read);
+        cache.mark_endpoint_unavailable(endpoint1, RequestOperation::Write);
+
+        assert_eq!(
+            cache
+                .location_unavailability_info_map
+                .lock()
+                .unwrap()
+                .get(endpoint1)
+                .map(|info| info.unavailable_operation),
+            Some(RequestOperation::All)
+        );
+    }
+
+    #[test]
+    fn refresh_stale_endpoints_test() {
+        // create test data
+        let (default_endpoint, write_locations, read_locations, preferred_locations) =
+            create_test_data();
+
+        let mut cache = LocationCache::new(default_endpoint, preferred_locations);
+
+        cache.update(write_locations, read_locations);
+
+        // mark endpoint 1 and endpoint 2 as unavailable
+        let endpoint1 = "https://location1.documents.azure.com";
+        let endpoint2 = "https://location2.documents.azure.com";
+        cache.mark_endpoint_unavailable(endpoint1, RequestOperation::Read);
+        cache.mark_endpoint_unavailable(endpoint2, RequestOperation::Read);
+
+        // simulate stale entry
```

Comment:
Great call! It's much better to forcibly inject a stale entry than to try to have the test "wait" until an entry goes stale.